core/lib/zksync_core/src/vm_runner/storage.rs

```diff
@@ -0,0 +1,400 @@
+                );
+                Ok(None)
+            }
+            Some(batch_data) => Ok(Some(batch_data.execute_data.clone())),
+        }
+    }
+}
+
+#[async_trait]
+impl<L: VmRunnerStorageLoader> ReadStorageFactory for VmRunnerStorage<L> {
+    async fn access_storage(
+        &self,
+        stop_receiver: &watch::Receiver<bool>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'_>>> {
+        self.access_storage_inner(stop_receiver, l1_batch_number)
+            .await
+    }
+}
+
+/// A runnable task that catches up the provided RocksDB cache instance to the latest processed
+/// batch and then continuously makes sure that this invariant is held for the foreseeable future.
+/// In the meanwhile, `StorageSyncTask` also loads the next `max_batches_to_load` batches in memory
+/// so that they are immediately accessible by [`VmRunnerStorage`].
+#[derive(Debug)]
+pub struct StorageSyncTask<L: VmRunnerStorageLoader> {
+    pool: ConnectionPool<Core>,
+    l1_batch_params_provider: L1BatchParamsProvider,
+    chain_id: L2ChainId,
+    rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+    loader: L,
+    state: Arc<RwLock<State>>,
+    catchup_task: AsyncCatchupTask,
+}
+
+impl<L: VmRunnerStorageLoader> StorageSyncTask<L> {
+    async fn new(
+        pool: ConnectionPool<Core>,
+        chain_id: L2ChainId,
+        rocksdb_path: String,
+        loader: L,
+        state: Arc<RwLock<State>>,
+    ) -> anyhow::Result<Self> {
+        let mut conn = pool.connection_tagged(L::name()).await?;
+        let l1_batch_params_provider = L1BatchParamsProvider::new(&mut conn)
+            .await
+            .context("Failed initializing L1 batch params provider")?;
+        let rocksdb_cell = Arc::new(OnceCell::new());
+        let catchup_task = AsyncCatchupTask::new(
+            pool.clone(),
+            rocksdb_path,
+            rocksdb_cell.clone(),
+            Some(loader.latest_processed_batch(&mut conn).await?),
+        );
+        drop(conn);
+        Ok(Self {
+            pool,
+            l1_batch_params_provider,
+            chain_id,
+            rocksdb_cell,
+            loader,
+            state,
+            catchup_task,
+        })
+    }
+
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        self.catchup_task.run(stop_receiver.clone()).await?;
+        let rocksdb = self.rocksdb_cell.get().ok_or_else(|| {
+            anyhow::anyhow!("Expected RocksDB to be initialized by `AsyncCatchupTask`")
+        })?;
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("`StorageSyncTask` was interrupted");
+                return Ok(());
+            }
+            // It's important to hold a lock on `state` while we are updating RocksDB cache
+            // as otherwise `VmRunnerStorage` will have an inconsistent view on DB state.
+            let mut state = self.state.write().await;
```

Comment:
hmm, if storage is used as intended, i.e. one never holds storage access to a batch with number `<= latest_processed_batch`, then it should be safe to write to RocksDB in parallel with data being read from it. I would prefer not to hold a lock here.
We can just say that implementation relies on the fact that one never uses storage pointer to a batch with number `<= latest_processed_batch`, otherwise -- UB

Reply:
Aren't both approaches pretty much equivalent to each other though?

If we expect the users to drop all storage access handles, it effectively means we expect them to drop all read locks on `state`. And they wouldn't be able to start reading again before we synchronize RocksDB to the next `latest_processed_batch ` anyway, effectively meaning they are waiting for write lock on `state` to drop.

Reply:
it seems no, e.g. if latest_processed_batch=10, last_ready_to_be_loaded_batch=12. Batches 11 and 12 are being processed. Then 11th becomes processed and I suggest that we should update rocksDB with storage logs from 11th batch while 12th is still in-progress. 
Moreover, the main loop of `StorageSyncTask` firstly synchronizes RocksDB and then loads data for new batches to memory, so with the current approach, when batch 13 becomes ready it cannot be started before 12 finishes which contradicts the purpose of vm runner storage

Reply:
You are right, I just had a brain fart that we can't update `latest_processed_batch` to anything other than `last_ready_to_be_loaded_batch`.

Reply:
Should be done now, PTAL. Also, we should only prohibit access storage handles for batches `< latest_processed_batch`. Equivalent are still okay and even necessary to process the next batch.